File: super_grasp/src/super_grasp/point_cloud_data.cpp
#include "super_grasp/point_cloud_data.h"
#include <chrono>

namespace super_grasp {

PointCloudData::PointCloudData(const sensor_msgs::ImageConstPtr& raw_rgb,
                               const sensor_msgs::ImageConstPtr& raw_depth,
                               const std::vector<geometry_msgs::Point>& mask,
                               const sensor_msgs::CameraInfoConstPtr& camera_info,
                               int nearest_neighbor,
                               double distance_thresh,
                               int segment_threshold,
                               bool debug)
    : mask_(mask),
      camera_info_(camera_info),
      nearest_neighbor_(nearest_neighbor),
      distance_thresh_(distance_thresh),
      segment_threshold_(segment_threshold),
      debug_(debug) {

    auto start_time = std::chrono::high_resolution_clock::now();

    // Convert ROS images to OpenCV
    try {
        cv_bridge::CvImageConstPtr cv_rgb = cv_bridge::toCvShare(raw_rgb, "bgr8");
        cv_bridge::CvImageConstPtr cv_depth = cv_bridge::toCvShare(raw_depth);
        raw_rgb_ = cv_rgb->image.clone();
        raw_depth_ = cv_depth->image.clone();
    } catch (cv_bridge::Exception& e) {
        ROS_ERROR("cv_bridge exception: %s", e.what());
        return;
    }

    // Convert to point cloud
    pcd_ = convertToPCD();

    if (debug_) {
        auto end_time = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);
        ROS_INFO("pointCloudData: Point cloud generated within %.3fs", duration.count() / 1000.0);
    }
}

std::shared_ptr<open3d::geometry::PointCloud> PointCloudData::removeOutliers(
    std::shared_ptr<open3d::geometry::PointCloud> pcd) {
    
    try {
        if (pcd->IsEmpty()) {
            ROS_WARN("pointCloudData [removeOutliers]: Provided point cloud is empty.");
            return pcd;
        }

        // Statistical outlier removal
        auto [cleaned_pcd, ind] = pcd->RemoveStatisticalOutliers(nearest_neighbor_, 0.25);
        
        return cleaned_pcd;
    } catch (const std::exception& e) {
        ROS_ERROR("pointCloudData [removeOutliers] Error: %s", e.what());
        return pcd;
    }
}

std::shared_ptr<open3d::geometry::PointCloud> PointCloudData::convertToPCD() {
    try {
        // Create mask image
        cv::Mat mask = cv::Mat::zeros(camera_info_->height, camera_info_->width, CV_8UC1);
        
        std::vector<cv::Point> points;
        for (const auto& pt : mask_) {
            points.push_back(cv::Point(static_cast<int>(pt.x), static_cast<int>(pt.y)));
        }
        
        cv::fillPoly(mask, points, cv::Scalar(255));

        // Apply mask to RGB and depth
        cv::Mat rgb_masked = cv::Mat::zeros(raw_rgb_.size(), raw_rgb_.type());
        cv::Mat depth_masked = cv::Mat::zeros(raw_depth_.size(), raw_depth_.type());
        
        raw_rgb_.copyTo(rgb_masked, mask);
        raw_depth_.copyTo(depth_masked, mask);

        // Ensure depth is in correct format (mm as uint16 or meters as float)
        cv::Mat depth_float;
        if (depth_masked.type() == CV_16UC1) {
            depth_masked.convertTo(depth_float, CV_32F, 1.0 / 1000.0); // Convert mm to meters
        } else if (depth_masked.type() == CV_32FC1) {
            depth_float = depth_masked;
        } else {
            depth_masked.convertTo(depth_float, CV_32F);
        }

        // Create Open3D images
        auto color_open3d = std::make_shared<open3d::geometry::Image>();
        auto depth_open3d = std::make_shared<open3d::geometry::Image>();
        
        color_open3d->Prepare(raw_rgb_.cols, raw_rgb_.rows, 3, 1);
        depth_open3d->Prepare(depth_float.cols, depth_float.rows, 1, 4);

        // Copy data
        memcpy(color_open3d->data_.data(), rgb_masked.data, 
               rgb_masked.total() * rgb_masked.elemSize());
        memcpy(depth_open3d->data_.data(), depth_float.data,
               depth_float.total() * depth_float.elemSize());

        // Create RGBD image
        auto rgbd = open3d::geometry::RGBDImage::CreateFromColorAndDepth(
            *color_open3d, *depth_open3d, 1000.0, 10.0, false);

        // Extract camera intrinsics
        double fx = camera_info_->K[0];
        double fy = camera_info_->K[4];
        double cx = camera_info_->K[2];
        double cy = camera_info_->K[5];
        int width = camera_info_->width;
        int height = camera_info_->height;

        open3d::camera::PinholeCameraIntrinsic intrinsic(width, height, fx, fy, cx, cy);

        // Generate point cloud
        auto point_cloud = open3d::geometry::PointCloud::CreateFromRGBDImage(*rgbd, intrinsic);

        if (debug_) {
            ROS_INFO("pointCloudData: pointcloud size %zu", point_cloud->points_.size());
        }

        // Remove outliers
        point_cloud = removeOutliers(point_cloud);

        return point_cloud;

    } catch (const std::exception& e) {
        ROS_ERROR("pointCloudData [convertToPCD] Error: %s", e.what());
        return std::make_shared<open3d::geometry::PointCloud>();
    }
}

std::vector<std::shared_ptr<open3d::geometry::PointCloud>> 
PointCloudData::defineSegments(std::shared_ptr<open3d::geometry::PointCloud> pcd) {
    
    try {
        auto start_time = std::chrono::high_resolution_clock::now();

        size_t remaining_points_threshold = pcd->points_.size() / segment_threshold_;
        std::vector<std::shared_ptr<open3d::geometry::PointCloud>> cloud_segments;
        
        auto remaining = std::make_shared<open3d::geometry::PointCloud>(*pcd);

        while (remaining->points_.size() > remaining_points_threshold) {
            // Segment plane using RANSAC
            auto [plane_model, inliers] = remaining->SegmentPlane(
                distance_thresh_, 3, 1000, 0.999);

            if (inliers.size() == 0) {
                break;
            }

            auto inlier_cloud = remaining->SelectByIndex(inliers);
            cloud_segments.push_back(inlier_cloud);

            remaining = remaining->SelectByIndex(inliers, true); // Invert selection
        }

        if (debug_) {
            auto end_time = std::chrono::high_resolution_clock::now();
            auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);
            ROS_INFO("pointCloudData: Segmented into %zu planes. Generated within %.3fs",
                     cloud_segments.size(), duration.count() / 1000.0);
        }

        return cloud_segments;

    } catch (const std::exception& e) {
        ROS_ERROR("pointCloudData [defineSegments] Error: %s", e.what());
        return std::vector<std::shared_ptr<open3d::geometry::PointCloud>>();
    }
}

std::vector<std::shared_ptr<open3d::geometry::PointCloud>> 
PointCloudData::getCloudSegments() {
    return defineSegments(pcd_);
}

} // namespace super_grasp

File: src/generate_grasp/pointCloudData.py
#!/usr/bin/env python3.8

"""
    pointCloudData: extracts data from raw rgb and depth input, 
                    applies a mask and removes outliers
                    also can segment the object model into multiple planes using RANSAC

    Input:
        raw_rgb: raw rgb image from camera
        raw_depth: raw depth image from camera
        mask: mask of the object to extract from rgb and depth image
        camera_info: camera info from camera
        nearest_neighbor: number of nearest neighbors to consider for outlier removal
        distance_thresh: distance threshold for RANSAC plane segmentation
        semght_threshold: threshold to stop segmentation based on remaining points
        debug: if True, print debug information

    Output:
        pcd: point cloud of the masked object
        cloud_segments: list of point cloud segments from RANSAC

    Developed by: Jayden Kanbour as part of undergraduate thesis for UNSW Computer Science and Engineering
    Date: 26th November 2025
    Email: jkanbour1@gmail.com
    UNSW Student Id: z5316799

"""

import cv2
import numpy as np
import open3d as o3d
import time
import copy
from cv_bridge import CvBridge

class PointCloudData:

    def __init__(self, raw_rgb, raw_depth, mask, camera_info, nearest_neighbor=500, distance_thresh=0.005, semght_threshold=10, debug=False):

        #convert ROS images to CV2
        self.bridge = CvBridge()
        self.raw_rgb = self.bridge.imgmsg_to_cv2(raw_rgb, 'bgr8')
        self.raw_depth = self.bridge.imgmsg_to_cv2(raw_depth, desired_encoding="passthrough")

        self.debug = debug
        self.mask = mask
        self.camera_info = camera_info
        self.nearest_neighbout = nearest_neighbor
        self.distance_thresh = distance_thresh
        self.segmentation_threshold = semght_threshold


        start_time = time.time()

        #convert extracted data to point cloud
        self.pcd = self.covertToPCD()

        if self.debug:
            print(f"pointCloudData: Point cloud generated within {time.time() - start_time:.3f}s")
    
    #remove outliers from, may not need if mask is good
    def removeOutliers(self, pcd):
        """
            Removes outliers based on nearest neighbour algorithm
        """
        try:
            if pcd.is_empty():
                print("pointCloudData [removeOutliers] Warning: Provided point cloud is empty.")
                return pcd
            
            # Efficient parameters
            _, ind = pcd.remove_statistical_outlier(nb_neighbors=self.nearest_neighbout, std_ratio=0.25)

            return pcd.select_by_index(ind)

        except Exception as e:
            print(f"pointCloudData [removeOutliers] Error: {e}")
            return None

    def covertToPCD(self):
        """
            Applied mask to masked RGB-D image and generated a point cloud object from the results
        """

        try:
            # Apply mask to RGB and Depth images
            mask = np.zeros((self.camera_info.height, self.camera_info.width), dtype=np.uint8)
            points = np.array(self.mask).reshape(-1, 2)
            mask = cv2.fillPoly(mask, [points.astype(np.int32)], 255)
            mask = mask.astype(bool) 
            rgb_masked = np.where(mask[:, :, None], self.raw_rgb, 0).astype(np.uint8)
            depth_masked = np.where(mask, self.raw_depth, 0)
            
            # Ensure depth is in uint16 (mm)
            if depth_masked.dtype != np.uint16:
                depth_masked = (depth_masked * 1000).astype(np.uint16)
            
            # Create Open3D RGBD image
            rgb_o3d = o3d.geometry.Image(rgb_masked)
            depth_o3d = o3d.geometry.Image(depth_masked)
            rgbd = o3d.geometry.RGBDImage.create_from_color_and_depth(
                color=rgb_o3d,
                depth=depth_o3d,
                depth_scale=1000.0,
                depth_trunc=10,
                convert_rgb_to_intensity=False
            )

            # Extract Camera intrinsics
            K = np.array(self.camera_info.K).reshape(3, 3)
            fx = K[0, 0]
            fy = K[1, 1]
            cx = K[0, 2]
            cy = K[1, 2]
            w = self.camera_info.width
            h = self.camera_info.height
            intrinsic = o3d.camera.PinholeCameraIntrinsic(width=w, height=h, fx=fx, fy=fy, cx=cx, cy=cy)
            
            # Generate Point Cloud with open3d
            pointCloud = o3d.geometry.PointCloud.create_from_rgbd_image(rgbd, intrinsic)

            if self.debug:
                print(f'pointCloudData: pointcloud size', len(pointCloud.points))

            # remove outliers based on nearest neighbour
            pointCloud = self.removeOutliers(pointCloud)

            return pointCloud
            
        except Exception as e:
            print(f"pointCloudData [covertToPCD] Error: {e}")
            return None
        
    def defineSegments(self, pcd):
        """
            Segments the point cloud into multiple planes using RANSAC
        """     
        try:   
            start_time = time.time()

            remaing_points_threshold = len(pcd.points) // self.segmentation_threshold
            cloud_segments = []
            remaining = copy.deepcopy(pcd)

            while len(remaining.points) > remaing_points_threshold:
                _, inliers = remaining.segment_plane(distance_threshold=self.distance_thresh,
                                                    ransac_n=3,
                                                    num_iterations=1000,
                                                    probability=0.999)
                
                inlier_cloud = remaining.select_by_index(inliers)

                cloud_segments.append(inlier_cloud)
                remaining = remaining.select_by_index(inliers, invert=True)

            if self.debug:
                print(f"pointCloudData: Segmented into {len(cloud_segments)} planes. Generated within {time.time() - start_time:.3f}s")

            return cloud_segments
        
        except Exception as e:
            print(f"pointCloudData [defineSegments] Error: {e}")
            return None

    def getPCD(self):
        return self.pcd

    def getCenter(self):
        return self.pcd.get_center()

    def getCloudSegments(self):
        return self.defineSegments(self.pcd)